statsmodels/tsa/statespace/tbats.py

```diff
@@ -0,0 +1,1430 @@
+                initial_state_cov_stationary
+            )
+        self.initialize_known(initial_state, initial_state_cov)
+
+    def filter(self, params, **kwargs):
+        kwargs.setdefault('results_class', TBATSResults)
+        kwargs.setdefault('results_wrapper_class',
+                          TBATSResultsWrapper)
+        return super(TBATSModel, self).filter(params, **kwargs)
+
+    def smooth(self, params, **kwargs):
+        kwargs.setdefault('results_class', TBATSResults)
+        kwargs.setdefault('results_wrapper_class',
+                          TBATSResultsWrapper)
+        return super(TBATSModel, self).smooth(params, **kwargs)
+
+    def loglike(self, params, *args, **kwargs):
+        """
+        Loglikelihood evaluation
+
+        Parameters
+        ----------
+        params : array_like
+            Array of parameters at which to evaluate the loglikelihood
+            function.
+        **kwargs
+            Additional keyword arguments to pass to the Kalman filter. See
+            `KalmanFilter.filter` for more details.
+
+        Notes
+        -----
+        [1]_ recommend maximizing the average likelihood to avoid scale issues;
+        this is done automatically by the base Model fit method.
+
+        References
+        ----------
+        .. [1] Koopman, Siem Jan, Neil Shephard, and Jurgen A. Doornik. 1999.
+           Statistical Algorithms for Models in State Space Using SsfPack 2.2.
+           Econometrics Journal 2 (1): 107-60. doi:10.1111/1368-423X.00023.
+
+        See Also
+        --------
+        update : modifies the internal state of the state space model to
+                 reflect new params
+        """
+        # We need to handle positional arguments in two ways, in case this was
+        # called by a Scipy optimization routine
+        if len(args) > 0:
+            argnames = ['transformed', 'complex_step']
+            # the fit() method will pass a dictionary
+            if isinstance(args[0], dict):
+                flags = args[0]
+            # otherwise, a user may have just used positional arguments...
+            else:
+                flags = dict(zip(argnames, args))
+            transformed = flags.get('transformed', True)
+            complex_step = flags.get('complex_step', True)
+
+            for name, value in flags.items():
+                if name in kwargs:
+                    raise TypeError("loglike() got multiple values for keyword"
+                                    " argument '%s'" % name)
+        else:
+            transformed = kwargs.pop('transformed', True)
+            complex_step = kwargs.pop('complex_step', True)
+
+        if not transformed:
+            params = self.transform_params(params)
+
+        self.update(params, transformed=True, complex_step=complex_step)
+
+        if complex_step:
+            kwargs['inversion_method'] = INVERT_UNIVARIATE | SOLVE_LU
+
+        sse = self.ssm.loglike(complex_step=complex_step, **kwargs)
```

Comment:
Can everything up to this point just be computed by the base class using `sse = super(TBATSModel, self).loglike(params, *args, **kwargs)` rather than copying it here?

Reply:
the lambda can be cached when updating parameter for caculating the loglikelihood.
I will make the change next time.